strum_macros/src/macros/enum_map.rs

```diff
@@ -0,0 +1,217 @@
+use proc_macro2::{Span, TokenStream};
+use quote::{format_ident, quote};
+use syn::{spanned::Spanned, Data, DeriveInput, Fields, Ident};
+
+use crate::helpers::{non_enum_error, HasStrumVariantProperties};
+
+pub fn enum_map_inner(ast: &DeriveInput) -> syn::Result<TokenStream> {
+    let name = &ast.ident;
+    let gen = &ast.generics;
+    let vis = &ast.vis;
+    let mut doc_comment = format!("A map over the variants of [{}]", name);
+
+    if gen.lifetimes().count() > 0 {
+        return Err(syn::Error::new(
+            Span::call_site(),
+            "This macro doesn't support enums with lifetimes.",
+        ));
+    }
+
+    let Data::Enum(data_enum) = &ast.data else {
+        return Err(non_enum_error())
+    };
+    let map_name = syn::parse_str::<Ident>(&format!("{}Map", name)).unwrap();
+
+    let variants = &data_enum.variants;
+
+    // the identifiers of each variant, in PascalCase
+    let mut pascal_idents = Vec::new();
+    // the identifiers of each struct field, in snake_case
+    let mut snake_idents = Vec::new();
+    // match arms in the form `MyEnumMap::Variant => &self.variant,`
+    let mut get_matches = Vec::new();
+    // match arms in the form `MyEnumMap::Variant => &mut self.variant,`
+    let mut get_matches_mut = Vec::new();
+    // match arms in the form `MyEnumMap::Variant => self.variant = new_value`
+    let mut set_matches = Vec::new();
+    // struct fields of the form `variant: func(MyEnum::Variant),*
+    let mut closure_fields = Vec::new();
+    // struct fields of the form `variant: func(MyEnum::Variant, self.variant),`
+    let mut transform_fields = Vec::new();
+
+    // identifiers for disabled variants
+    let mut disabled_variants = Vec::new();
+    // match arms for disabled variants
+    let mut disabled_matches = Vec::new();
+
+    for variant in variants {
+        // skip disabled variants
+        if variant.get_variant_properties()?.disabled.is_some() {
+            let disabled_ident = &variant.ident;
+            let panic_message =
+                format!("Can't use `{disabled_ident}` with `{map_name}` - variant is disabled for Strum features");
+            disabled_variants.push(disabled_ident);
+            disabled_matches.push(quote!(#name::#disabled_ident => panic!(#panic_message),));
+            continue;
+        }
+
+        // Error on fields with data
+        let Fields::Unit = &variant.fields else {
+            return Err(syn::Error::new(
+                variant.fields.span(),
+                "This macro doesn't support enums with non-unit variants",
+            ))
+        };
+
+        let pascal_case = &variant.ident;
+        // switch PascalCase to snake_case. This naively assumes they use PascalCase
+        let snake_case = format_ident!(
```

Comment:
Perhaps for this code we could reuse `enum_is.rs`'s `snakify` since that already accomplishes the purpose?

Reply:
Yes, I just copied that from you